internal/webhook/external/handlers/refresh_warehouse.go

```diff
@@ -0,0 +1,101 @@
+package handlers
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/api"
+	"github.com/akuity/kargo/internal/indexer"
+	"github.com/akuity/kargo/internal/logging"
+	"github.com/akuity/kargo/internal/webhook/external/providers"
+)
+
+// NewRefreshWarehouseWebhook handles push events for the designated provider.
+// After the provider has been resolved and the request has been authenticated,
+// the kubeclient is queried for all warehouses that contain a subscription
+// to the repo in question. Those warehouses are then patched with a special
+// annotation that signals down stream logic to refresh the warehouse.
+func NewRefreshWarehouseWebhook(p providers.Provider, c client.Client) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logger := logging.LoggerFromContext(r.Context())
+		logger.Debug("authenticating request")
+		if err := p.Authenticate(r); err != nil {
+			logger.Error(err, "failed to authenticate request")
+			http.Error(w,
+				"failed to authenticate request",
+				http.StatusUnauthorized,
+			)
+			return
+		}
+		logger.Debug("request authenticated")
+
+		logger.Debug("retrieving source repo")
+		repo, err := p.Repository(r)
+		if err != nil {
+			logger.Error(err, "failed to retrieve source repo")
+			http.Error(w,
+				err.Error(),
+				http.StatusBadRequest,
+			)
+			return
+		}
+
+		logger.Info("repo retrieved", "name", repo)
+
+		ctx := r.Context()
+		var warehouses v1alpha1.WarehouseList
+		err = c.List(
+			ctx,
+			&warehouses,
+			client.MatchingFields{
+				indexer.WarehousesBySubscribedURLsField: repo,
+			},
+		)
+		if err != nil {
+			logger.Error(err, "failed to list warehouses")
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		logger.Debug("listed warehouses",
+			"num-warehouses", len(warehouses.Items),
+		)
+
+		resp := &struct {
+			WarehousesSuccessfullyRefreshed int               `json:"warehouses_successfully_refreshed"`
+			WarehousesFailedToRefresh       int               `json:"warehouses_failed_to_refresh"`
+			Errors                          map[string]string `json:"errors,omitempty"`
+		}{
+			Errors: make(map[string]string),
+		}
```

Comment:
GH won't do anything with this. I think it's fine for now, but I think we should drop this before we're done.

Reply:
Github displays these responses in their UI in webhook settings under `recent deliveries`. Do you think that could be useful?

Reply:
Yes and no...

Seeing _something_ is helpful, but I don't think we need this much structure, for a few reasons:

- In the event of success, you're never going to look at `WarehousesSuccessfullyRefreshed`.

- In the event of a failure, details of the failure shouldn't be divulged in the response. i.e. The `Errors` field is information we really shouldn't include.

If all we care about is knowing that one or more Warehouses didn't refresh, without the details of _why_ each did not refresh, then all that's required to convey that is a single, vague error message along the lines of, "Failed to refresh n of m Warehouses."